cmd/malwareScan.go

```diff
@@ -0,0 +1,166 @@
+package cmd
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"github.com/SAP/jenkins-library/pkg/command"
+	piperhttp "github.com/SAP/jenkins-library/pkg/http"
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/SAP/jenkins-library/pkg/telemetry"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+)
+
+var open = _open
+var getSHA256 = _getSHA256
+
+func _open(path string) (io.ReadCloser, error) {
+	return os.Open(path)
+}
+
+type malwareScanResponse struct {
+	MalwareDetected          bool
+	EncryptedContentDetected bool
+	ScanSize                 int
+	MimeType                 string
+	SHA256                   string
+}
+
+func malwareScan(config malwareScanOptions, telemetryData *telemetry.CustomData) {
+	// for command execution use Command
+	c := command.Command{}
+	// reroute command output to logging framework
+	c.Stdout(log.Entry().Writer())
+	c.Stderr(log.Entry().Writer())
+
+	// for http calls import  piperhttp "github.com/SAP/jenkins-library/pkg/http"
+	// and use a  &piperhttp.Client{} in a custom system
+	// Example: step checkmarxExecuteScan.go
+
+	httpClient := &piperhttp.Client{}
+
+	// error situations should stop execution through log.Entry().Fatal() call which leads to an os.Exit(1) in the end
+	err := runMalwareScan(&config, telemetryData, &c, httpClient)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+
+func runMalwareScan(config *malwareScanOptions, telemetryData *telemetry.CustomData, command execRunner,
+	httpClient piperhttp.Sender) error {
+
+	log.Entry().WithField("LogField", "Log field content").Infof("Scanning file \"%s\" for malware using service \"%s\"", config.File, config.Host)
+
+	var candidate io.ReadCloser
+	var err error
+
+	candidate, err = open(config.File)
+	if err != nil {
+		return err
+	}
+	defer candidate.Close()
+
+	opts := piperhttp.ClientOptions{Username: config.User, Password: config.Password}
+	httpClient.SetOptions(opts)
+
+	// piper http utils mashall some http response codes into errors. We wan't to check the status code
+	// ourselvs hence we wait with returning that error (maybe also related to errors others than http status codes)
+	response, httpErr := httpClient.SendRequest("POST", config.Host+"/scan", candidate, prepareHeaders(), nil)
+
+	defer func() {
```

Comment:
if httpErr!=nil the response could be nil. wouldn't the defered func() fail?

Reply:
fixed